File: frontend/src/types/analytics.ts
/**
 * Analytics types for `GET /campaigns/{id}/analytics`.
 *
 * Two rules the backend guarantees, which the UI is built on:
 * `interaction_rate` is `0` when `views` is `0`, and `by_option` carries a row
 * for every option including zero-click ones - so a chart never has to guess
 * at a missing bar.
 */

import type { CampaignObjective, OptionIntent, PrimaryMetric } from "./campaign";

export interface OptionBreakdown {
  option_id: string;
  position: number;
  key: string;
  label: string;
  intent: OptionIntent;
  clicks: number;
  /** Fraction of interactions, `0.0-1.0`. Sums to 1, or 0 when there are none. */
  share: number;
}

export interface TimeseriesPoint {
  /** `YYYY-MM-DD`. */
  date: string;
  views: number;
  interactions: number;
}

export interface CampaignAnalytics {
  campaign_id: string;
  objective: CampaignObjective;

  views: number;
  interactions: number;
  interaction_rate: number;
  unique_viewers: number;

  by_option: OptionBreakdown[];
  primary_metric: PrimaryMetric | null;

  first_activity_at: string | null;
  last_activity_at: string | null;
  timeseries: TimeseriesPoint[];
}

File: backend/src/schemas/analytics.py
"""Analytics response schemas."""

from datetime import date, datetime

from pydantic import Field

from src.schemas.common import PrimaryMetric, StrictModel
from src.schemas.enums import OptionIntent


class OptionBreakdown(StrictModel):
    """Per-option click counts.

    A row is returned for every option, including zero-click ones - a chart
    with a missing bar is a bug the frontend should not have to guess around.
    """

    option_id: str
    position: int
    key: str
    label: str
    intent: OptionIntent
    clicks: int = 0
    share: float = Field(
        0.0, ge=0.0, le=1.0, description="Fraction of interactions. Sums to 1.0, or 0 when none."
    )


class TimeseriesPoint(StrictModel):
    date: date
    views: int = 0
    interactions: int = 0


class CampaignAnalytics(StrictModel):
    campaign_id: str
    objective: str

    views: int = 0
    interactions: int = 0
    interaction_rate: float = Field(
        0.0, ge=0.0, description="interactions / views. 0 when views is 0 - never divides by zero."
    )
    unique_viewers: int = 0

    by_option: list[OptionBreakdown] = Field(default_factory=list)
    primary_metric: PrimaryMetric | None = None

    first_activity_at: datetime | None = None
    last_activity_at: datetime | None = None
    timeseries: list[TimeseriesPoint] = Field(default_factory=list)
